File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cstring>

using namespace std;

vector<int> graph[101];
pair<int,int> except[101];
int groupCnt;
int visitCnt[2];
int answer = 80808080;

bool visited[101];

bool exceptCheck(int exceptIdx, int cur, int next){
    int from = except[exceptIdx].first;
    int to = except[exceptIdx].second;
    if((from == cur && to == next) || (to == cur && from == next)){
        return true;
    }
    return false;
}

void dfs(int exceptIdx, int cur){
    visited[cur] = true;
    visitCnt[groupCnt]++;
    for(int i=0; i<graph[cur].size(); i++){
        int next = graph[cur][i];
        if(exceptCheck(exceptIdx, cur, next)) continue;
        if(visited[next]) continue;
        dfs(exceptIdx, next);
    }
}

int solution(int n, vector<vector<int>> wires) {
    int k = 0;
    for(auto wire : wires){
        graph[wire[0]].push_back(wire[1]);
        graph[wire[1]].push_back(wire[0]);
        
        except[k++] = {wire[0] , wire[1]};
    }
    for(int i=0; i<wires.size(); i++){
        for(int j=1; j<=n; j++){
            if(visited[j]) continue;
            dfs(i,j);
            groupCnt++;
        }
        if(groupCnt == 2){
            answer = min(answer, abs(visitCnt[0] - visitCnt[1]));
        }
        groupCnt = 0;
        visitCnt[0] = visitCnt[1] = 0;
        memset(visited, false, sizeof(visited));
    }
    
    return answer;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    
    static List<Integer>[] graph;
    static boolean[] visited;
    static int[] visitedCnt;
    
    
    public int solution(int n, int[][] wires) {
        // n : 송전탑 개수, wires : 전선 정보
        // 전선 중 하나를 끊어 송전탑 개수가 비슷하도록 해야함 (절대값 반환)
        
        graph = new ArrayList[n+1];
        
        int min = Integer.MAX_VALUE;
        
        for(int i=1; i<=n; i++) {
            graph[i] = new ArrayList<>();
        }
        
        for(int i= 0; i<wires.length; i++) {
            graph[wires[i][0]].add(wires[i][1]);
            graph[wires[i][1]].add(wires[i][0]);
        }
        
        for(int i = 0; i< wires.length; i++) {
            int v1 = wires[i][0];
            int v2 = wires[i][1];
            
            visited = new boolean[n+1];
            
            graph[v1].remove(Integer.valueOf(v2));
            graph[v2].remove(Integer.valueOf(v1));
            
            int cnt = dfs(1);
            
            int diff = Math.abs(cnt - (n-cnt));
            min = Math.min(min, diff);
            
            graph[v1].add(v2);
            graph[v2].add(v1);
        }
        
        return min;
    }
    
    static public int dfs(int start) {
        visited[start] = true;
        int cnt = 1;
        
        for(Integer next : graph[start]) {
            if(!visited[next]) {
                cnt += dfs(next);
            }
        }
        return cnt;
    }
}